drivers/dma/dma_silabs_siwx917.c

```diff
@@ -0,0 +1,634 @@
+               memory-memory (0)\n \
+               memory-peripheral (1)\n \
+               peripheral-memory (2)\n");
+		return -1;
+	}
+	return 0;
+}
+
+/*Function to validate and convert channel data width for sl layer usage*/
+static inline int siwx917_dma_data_width(uint32_t data_width)
+{
+	switch (data_width) {
+	case 1:
+		return SL_TRANSFER_SIZE_8; // 8-bit width
+	case 2:
+		return SL_TRANSFER_SIZE_16; // 16-bit width
+	case 4:
+		return SL_TRANSFER_SIZE_32; // 32-bit width
+	default:
+		printf("Invalid Source/Destination data size - must be,\n \
+               8-bit (1)\n \
+               16-bit (2)\n \
+               32-bit (4)\n");
+		return -1;
+	}
+	return 0;
+}
+
+/*Function to validate and convert channel burst length for sl layer usage*/
+static inline int siwx917_dma_burst_length(uint32_t blen)
+{
+	switch (blen) {
+	case 1:
+		return VALID_BURST_LENGTH; // 8-bit burst
+	default:
+		printf("Invalid Source/Destination burst length - must be,\n \
+               8-bit Increment(1)\n");
+		return -1;
+	}
+	return 0;
+}
+
+/*Function to validate and convert channel addr increment value for sl layer
+ * usage*/
+static inline int siwx917_dma_addr_adjustment(uint32_t adjustment)
+{
+	switch (adjustment) {
+	case 0:
+		break; // Addr Increment
+	case 2:
+		return UDMA_ADDR_INC_NONE; // No Address increment
+	default:
+		printf("Invalid Source/Destination Addr adjustment - must be,\n \
+               Increment (0)\n \
+               No change (2)\n");
+		return -1;
+	}
+	return 0;
+}
+
+/*Function to configure UDMA channel for transfer*/
+static int dma_siwx917_configure(const struct device *dev, uint32_t channel,
+				 struct dma_config *config)
+{
+
+	// Fetch DMA config and data structures from device.
+	const struct dma_siwx917_config *cfg = dev->config;
+	struct dma_siwx917_data *data = dev->data;
+
+	uint32_t dma_instance;
+	sl_status_t status = SL_STATUS_OK;
+	UDMA_RESOURCES UDMA_Resources;
+	static UDMA_Channel_Info channel_info[32]; // channel info structure to temporarily hold the
+						   // channel data.
+	uint32_t rsi_channel = channel - 1;
+
+	if (cfg->reg == UDMA0) {
+		dma_instance = UDMA0_INSTANCE; // udma0
+		// Expecting a fixed channel number between 1-32
+		if ((channel > CHANNEL_32) || (channel == 0)) {
+			return -EINVAL;
+		} else {
+			UDMA_Resources.udma_irq_num = UDMA0_IRQn;
+		}
+	} else if (cfg->reg == UDMA1) {
+		dma_instance = ULP_UDMA_INSTANCE; // ulp_udma
+		// Expecting a fixed channel number between 1-12
+		if ((channel > CHANNEL_12) || (channel == 0)) {
+			return -EINVAL;
+		} else {
+			UDMA_Resources.udma_irq_num = UDMA1_IRQn;
+		}
+	} else {
+		return -EINVAL;
+	}
+
+	UDMA_Resources.reg = cfg->reg; // UDMA register base address
+	UDMA_Resources.desc =
+		(RSI_UDMA_DESC_T *)
+			cfg->reg->CTRL_BASE_PTR; // SRAM address where UDMA descriptor is stored
+
+	// Disable the channel before configuring
+	if (sl_si91x_dma_channel_disable(dma_instance, channel) != SL_STATUS_OK) {
```

Comment:
Zephyr doesn't play well with `dma_instance`. I suggest to drop `sl_si91x_dma_channel_disable()` and directly call `RSI_UDMA_ChannelDisable()`. So, you could get rid of `dma_instance`.

Reply:
For this case, I added RSI_UDMA_ChannelDisable() and tried to avoid using dma_instance wherever possible.